### PR TITLE
Improve the scheduling algorithm

### PR DESCRIPTION
**Pull request by BoxiLi**
- Improve the readability of the scheduling algorithm.
- Add more commutation rules, such as CNOT and X on the targets, CNOT and Z on the control.

Most of the changes in the test file are just because of black styling.

@nathanshammah @quantshah

### CODE REVIEW

File: tests/test_scheduler.py [hunk 1]
```diff
@@ -7,6 +7,62 @@
 from qutip import process_fidelity, qeye, tracedist
 
 
+def _varify_scheduled_circuit(circuit, gate_cycle_indices):
```
Comment (nathanshammah) [inline tests/test_scheduler.py:10]:
```suggestion
def _verify_scheduled_circuit(circuit, gate_cycle_indices):
```

File: tests/test_scheduler.py [hunk 2]
```diff
@@ -7,6 +7,62 @@
...
+    [
+        pytest.param(_circuit_cnot_x(), 2, id="cnot x commutation"),
+        pytest.param(_circuit_cnot_z(), 2, id="cnot z commutation"),
+    ],
+)
+def test_commutation_rules(circuit, expected_length):
+    scheduler = Scheduler("ASAP")
+    gate_cycle_indices = scheduler.schedule(circuit)
+    assert (max(gate_cycle_indices) + 1) == expected_length
+    assert _varify_scheduled_circuit(circuit, gate_cycle_indices)
```
Comment (nathanshammah) [inline tests/test_scheduler.py:63]:
```suggestion
    assert _verify_scheduled_circuit(circuit, gate_cycle_indices)
```

File: tests/test_scheduler.py [hunk 3]
```diff
@@ -74,18 +138,11 @@ def test_scheduling_gates1(
...
-    scheduled_gate = [[] for i in range(max(gate_cycle_indices)+1)]
-
-    # check if the scheduled circuit is correct
-    for i, cycles in enumerate(gate_cycle_indices):
-        scheduled_gate[cycles].append(circuit.gates[i])
-    circuit.gates = sum(scheduled_gate, [])
-    result1 = gate_sequence_product(circuit.propagators())
-    assert(tracedist(result0*result1.dag(), qeye(result0.dims[0])) < 1.0e-7)
+    assert max(gate_cycle_indices) == expected_length
+    _varify_scheduled_circuit(circuit, gate_cycle_indices)
```
Comment (nathanshammah) [inline tests/test_scheduler.py:145]:
```suggestion
    _verify_scheduled_circuit(circuit, gate_cycle_indices)
```

File: tests/test_scheduler.py [hunk 4]
```diff
@@ -107,26 +168,29 @@ def test_scheduling_gates2(
...
     gate_cycle_indices = scheduler.schedule(
-        circuit, gates_schedule=gates_schedule, repeat_num=repeat_num)
+        circuit, gates_schedule=gates_schedule, repeat_num=repeat_num
+    )
 
-    # check if the scheduled length is expected
-    assert(max(gate_cycle_indices) == expected_length)
-    scheduled_gate = [[] for i in range(max(gate_cycle_indices)+1)]
+    assert max(gate_cycle_indices) == expected_length
+    _varify_scheduled_circuit(circuit, gate_cycle_indices)
```
Comment (nathanshammah) [inline tests/test_scheduler.py:175]:
```suggestion
    _verify_scheduled_circuit(circuit, gate_cycle_indices)
```

File: tests/test_scheduler.py [hunk 5]
```diff
@@ -136,27 +200,29 @@ def test_scheduling_gates3(
...
+    )
 
-    # check if the scheduled circuit is correct
-    for i, cycles in enumerate(gate_cycle_indices):
-        scheduled_gate[cycles].append(circuit.gates[i])
-    circuit.gates = sum(scheduled_gate, [])
-    result1 = gate_sequence_product(circuit.propagators())
-    assert(tracedist(result0*result1.dag(), qeye(result0.dims[0])) < 1.0e-7)
+    assert max(gate_cycle_indices) == expected_length
+    _varify_scheduled_circuit(circuit, gate_cycle_indices)
```
Comment (nathanshammah) [inline tests/test_scheduler.py:207]:
```suggestion
    _verify_scheduled_circuit(circuit, gate_cycle_indices)
```

File: tests/test_scheduler.py [hunk 6]
```diff
@@ -166,34 +232,42 @@ def test_scheduling_gates4(
...
-    scheduled_gate = [[] for i in range(max(gate_cycle_indices)+1)]
-
-    # check if the scheduled circuit is correct
-    for i, cycles in enumerate(gate_cycle_indices):
-        scheduled_gate[cycles].append(circuit.gates[i])
-    circuit.gates = sum(scheduled_gate, [])
-    result1 = gate_sequence_product(circuit.propagators())
-    assert(tracedist(result0*result1.dag(), qeye(result0.dims[0])) < 1.0e-7)
+    assert max(gate_cycle_indices) == expected_length
+    _varify_scheduled_circuit(circuit, gate_cycle_indices)
```
Comment (nathanshammah) [inline tests/test_scheduler.py:239]:
```suggestion
    _verify_scheduled_circuit(circuit, gate_cycle_indices)
```

**Review by nathanshammah — CHANGES_REQUESTED** [review]
Thanks @BoxiLi. 

There is a small typo in the definition of a private function, please commit the suggestions. 

Can you please add more docstrings  in the private functions? Doesn't have to be completely in Numpy style, but more context would be good, even if the function names are quite pythonic.  

Also, are these features included in the documentation (API-doc)? Shall we add something about this?

**Comment by BoxiLi** [comment]
It is described in the API-doc, currently, it is not included in the user guide though.

**Comment by nathanshammah** [comment]
Is this ready for review again @BoxiLi?

**Review by nathanshammah — APPROVED** [review]
@BoxiLi looks fine to me. 

Very small comment: can we find some more "pythonic" name than `next_gate_ind`? `next_gate_index`? If not applicable, feel free to merge.

**Pull request merged**